File: terminal_ticker/market_data/router.py
"""文件用途：数据源路由，按配置把标的分发给对应 provider。"""
from __future__ import annotations

from . import alpaca, bitget, longbridge
from .alpaca import AlpacaInstrument
from .bitget import BitgetInstrument
from ..config import ALPACA_SOURCE, BITGET_SOURCE, InstrumentConfig, LONGBRIDGE_SOURCE
from .longbridge import LongbridgeInstrument

MarketInstrument = AlpacaInstrument | BitgetInstrument | LongbridgeInstrument


def resolve_instruments(configured: tuple[InstrumentConfig, ...]) -> tuple[MarketInstrument, ...]:
    # Resolve by provider, then rebuild the original watchlist order for the UI.
    """说明：把配置标的解析为具体 provider 标的，并保持 watchlist 顺序。"""
    bitget_configs = tuple(item for item in configured if item.source == BITGET_SOURCE)
    alpaca_configs = tuple(item for item in configured if item.source == ALPACA_SOURCE)
    longbridge_configs = tuple(item for item in configured if item.source == LONGBRIDGE_SOURCE)

    resolved_bitget = {}
    if bitget_configs:
        resolved_bitget = {
            config.dedupe_key: instrument
            for config, instrument in zip(bitget_configs, bitget.resolve_instruments(bitget_configs))
        }

    resolved_alpaca = {}
    if alpaca_configs:
        resolved_alpaca = {
            config.dedupe_key: instrument
            for config, instrument in zip(alpaca_configs, alpaca.resolve_instruments(alpaca_configs))
        }

    resolved_longbridge = {}
    if longbridge_configs:
        resolved_longbridge = {
            config.dedupe_key: instrument
            for config, instrument in zip(
                longbridge_configs,
                longbridge.resolve_instruments(longbridge_configs),
            )
        }

    resolved: list[MarketInstrument] = []
    for config in configured:
        if config.source == BITGET_SOURCE:
            resolved.append(resolved_bitget[config.dedupe_key])
        elif config.source == ALPACA_SOURCE:
            resolved.append(resolved_alpaca[config.dedupe_key])
        elif config.source == LONGBRIDGE_SOURCE:
            resolved.append(resolved_longbridge[config.dedupe_key])
        else:
            raise ValueError(f"unsupported data source: {config.source}")
    return tuple(resolved)

File: tradex/market_data/router.ts
import { BITGET_SOURCE, InstrumentConfig } from "../config/index.js";
import { BitgetInstrument, resolveInstruments as resolveBitgetInstruments } from "./bitget.js";

export type MarketInstrument = BitgetInstrument;

function dedupeKey(config: InstrumentConfig): string {
  return `${config.source}\0${config.instType ?? ""}\0${config.symbol}`;
}

export async function resolveInstruments(configured: readonly InstrumentConfig[]): Promise<MarketInstrument[]> {
  const bitgetConfigs = configured.filter((item) => item.source === BITGET_SOURCE);
  const resolvedBitget = new Map<string, BitgetInstrument>();
  if (bitgetConfigs.length > 0) {
    const instruments = await resolveBitgetInstruments(bitgetConfigs);
    bitgetConfigs.forEach((config, index) => resolvedBitget.set(dedupeKey(config), instruments[index]));
  }
  return configured.map((config) => {
    if (config.source === BITGET_SOURCE) {
      const instrument = resolvedBitget.get(dedupeKey(config));
      if (!instrument) throw new Error(`unresolved Bitget instrument: ${config.symbol}`);
      return instrument;
    }
    throw new Error(`unsupported data source: ${config.source}`);
  });
}
